packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts

```diff
@@ -0,0 +1,530 @@
+    }
+    this.maxEpochsInMemory = opts.maxEpochsInMemory;
+    // Specify different persistentApis for testing
+    this.persistentApis = persistentApis;
+    this.shufflingCache = shufflingCache;
+    this.getHeadState = getHeadState;
+    this.inMemoryEpochs = new Set();
+  }
+
+  /**
+   * Get a state from cache, it will reload from disk.
+   * This is expensive api, should only be called in some important flows:
+   * - Validate a gossip block
+   * - Get block for processing
+   * - Regen head state
+   */
+  async getOrReload(cp: CheckpointHex): Promise<CachedBeaconStateAllForks | null> {
+    const cpKey = toCheckpointKey(cp);
+    const inMemoryState = this.get(cpKey);
+    if (inMemoryState) {
+      return inMemoryState;
+    }
+
+    const persistentKey = this.cache.get(cpKey);
+    if (persistentKey === undefined) {
+      return null;
+    }
+
+    if (!isPersistentKey(persistentKey)) {
+      // should not happen, in-memory state is handled above
+      throw new Error("Expected persistent key");
+    }
+
+    // reload from disk or db based on closest checkpoint
+    this.logger.verbose("Reload: read state", {persistentKey});
+    const newStateBytes = await this.persistentApis.read(persistentKey);
+    if (newStateBytes === null) {
+      this.logger.verbose("Reload: read state failed", {persistentKey});
+      return null;
+    }
+    this.logger.verbose("Reload: read state successfully", {persistentKey});
+    this.metrics?.stateRemoveCount.inc({reason: RemovePersistedStateReason.reload});
+    this.metrics?.stateReloadSecFromSlot.observe(this.clock?.secFromSlot(this.clock?.currentSlot ?? 0) ?? 0);
+    const closestState = findClosestCheckpointState(cp, this.cache) ?? this.getHeadState?.();
+    if (closestState == null) {
+      throw new Error("No closest state found for cp " + toCheckpointKey(cp));
+    }
+    this.metrics?.stateReloadEpochDiff.observe(Math.abs(closestState.epochCtx.epoch - cp.epoch));
+    this.logger.verbose("Reload: found closest state", {persistentKey, seedSlot: closestState.slot});
+    const timer = this.metrics?.stateReloadDuration.startTimer();
+
+    try {
+      const newCachedState = loadCachedBeaconState(closestState, newStateBytes, {
+        shufflingGetter: this.shufflingCache.get.bind(this.shufflingCache),
+      });
+      timer?.();
+      this.logger.verbose("Reload state successfully", {
+        persistentKey,
+        stateSlot: newCachedState.slot,
+        seedSlot: closestState.slot,
+      });
+      // only remove persisted state once we reload successfully
+      void this.persistentApis.remove(persistentKey);
+      this.cache.set(cpKey, newCachedState);
+      this.inMemoryEpochs.add(cp.epoch);
+      // don't prune from memory here, call it at the last 1/3 of slot 0 of an epoch
+      return newCachedState;
+    } catch (e) {
+      this.logger.debug("Error reloading state from disk", {persistentKey}, e as Error);
+      return null;
```

Comment:
Why try catch and return null? What's the API contract of this function then?

Reply:
If cannot find an in-memory state or unable to reload, return null. Will update the method header.